src/zc/lockfile/__init__.py

```diff
@@ -102,3 +103,61 @@ def close(self):
             _unlock_file(self._fp)
             self._fp.close()
             self._fp = None
+
+
+class LockedContext(object):
+    """ContextManager version of LockFile"""
+
+    def __init__(self, name, content_template='{pid}', timeout=float("inf")):
+        self.name = str(name)  # wrap in str() to allow pathlib usage
+        self.content_template = content_template
+        self.lock = None
+        self.timeout = timeout
+
+    def __enter__(self):
```

Comment:
What if the timeout and retry logic was made part of the primitive operations in `LockFile`? If you do that (and implement `open_exclusive` through composition), then the context manager becomes pretty simple:

```python
from contextlib import closing
@contextmanager
def locked_file(*lock_args, **lock_kwargs):
    with closing(LockFile(*lock_args, **lock_kwargs)) as lock:
        yield lock
```

Reply:
Given https://github.com/zopefoundation/zc.lockfile/pull/16/ we'll probably need to revisit this and make the class of the lock configurable...or implement the context manager as part of the base class.

Reply:
> Given #16 we'll probably need to revisit this and make the class of the lock configurable...or implement the context manager as part of the base class.

For example, we might define a new class `Lock` that implements the primitive `acquire(blocking,timeout)` and `release()` methods. That class would have the hooks that #16 is asking for.

For backwards compatibility, `LockFile` would still exist and just implement `__init__` and `close()` in terms of `Lock().acquire()` and `Lock().release()`. 

`Lock` would itself be a context manager. So someone that has a customized acquisition behaviour would write `with MyCustomLock(...)` (or perhaps they wouldn't even need a new class, they could just use `@contextmanager` around a function that creates a `SimpleLock`). 

Then `open_exclusive` would be written:
```python
@contextmanager
def open_exclusive(self, filename, lock_file=None, **open_args):
    if not isinstance(lock_file, Lock):
        # Pass either a path name or a lock file
        lock_file_name = filename + os.path.extsep + 'lock'
        lock_file = Lock(lock_file_name)

    with lock_file:
        with open(filename, **open_args) as f:
               yield f
```